prov/efa/src/rdm/rxr_pkt_type_base.c

```diff
@@ -377,39 +377,59 @@ int rxr_pkt_copy_data_to_cuda(struct rxr_ep *ep,
 	static const int max_blocking_copy_rx_entry_num = 4;
 	struct rxr_op_entry *rx_entry;
 	struct efa_mr *desc;
-	bool p2p_available, blocking_copy_available;
+	bool p2p_available, local_read_available, gdrcopy_available, cuda_memcpy_available;
 	int ret, err;
 
 	rx_entry = pkt_entry->x_entry;
 	desc = rx_entry->desc[0];
 	assert(efa_mr_is_cuda(desc));
 
-	blocking_copy_available = cuda_is_gdrcopy_enabled() || (cuda_get_xfer_setting() == CUDA_XFER_ENABLED);
+	local_read_available = false;
+	gdrcopy_available = false;
+	cuda_memcpy_available = false;
+
+	if (cuda_get_xfer_setting() == CUDA_XFER_ENABLED) {
+		cuda_memcpy_available = true;
+	} else {
+		/* gdrcopy is not enabled when CUDA API is allowed to be used */
+		gdrcopy_available = cuda_is_gdrcopy_enabled();
```

Comment:
This `if... else...` block is dubious to me.

Intuitively `gdrcopy_available` should always be `cuda_is_gdrcopy_enabled()`. The `if... else...` implies a mutual exclusion, i.e. `cuda_memcpy_available` and `gdrcopy_available` cannot both be true.

Symmetry would be better IMO, i.e. extract a new function `cuda_is_memcpy_enabled()`, then replace
```c
        if (cuda_get_xfer_setting() == CUDA_XFER_ENABLED) {
		cuda_memcpy_available = true;
	} else {
		/* gdrcopy is not enabled when CUDA API is allowed to be used */
		gdrcopy_available = cuda_is_gdrcopy_enabled();
        {
```
with 
```c
        cuda_memcpy_available = cuda_is_memcpy_enabled();
        gdrcopy_available = cuda_is_gdrcopy_enabled();
```

Reply:
Your understanding is correct. `cuda_memcpy_available` and `gdrcopy_available` are mutually exclusive. As the comment stated "gdrcopy is not enabled when CUDA API is allowed to be used".

Maybe I should rephrase to that "When CUDA API is available, efa provider will not use gdrcopy even if it is enabled". (Would that be move clear to you)?

Reply:
Yes I think the new comment is more clear and describes the actual intention.

Reply:
Fixed in new revision.

Reply:
I think the difference is "available" vs "preferred".  I think GDR and CudaMemCopy can both be available, but the logic here is trying to decide which is preferred.

Reply:
>  I think GDR and CudaMemCopy can both be available

No they cannot, because gdrcopy require register memory with the gdrdrv kernel module, and we do not register memory with gdrcopy when cuda API is available to be used. See. https://github.com/ofiwg/libfabric/blob/76c449a93e482e8c2bd2ab7a85c29c6067bc9d75/prov/efa/src/efa_mr.c#L213 and https://github.com/ofiwg/libfabric/blob/76c449a93e482e8c2bd2ab7a85c29c6067bc9d75/prov/efa/src/efa_mr.c#L520

Reply:
In new revision, I added a link to the function `#efa_mr_use_gdrcopy` for reader's reference.